pylgbst/comms/cbluepy.py

```diff
@@ -0,0 +1,123 @@
+    if not m:
+        raise ValueError('Cannot find iface number in {}.'.format(controller))
+    return int(m.group(1))
+
+
+class BluepyDelegate(btle.DefaultDelegate):
+    def __init__(self, handler):
+        btle.DefaultDelegate.__init__(self)
+
+        self._handler = handler
+
+    def handleNotification(self, cHandle, data): 
+        log.debug('Incoming notification')
+        self._handler(cHandle, data)
+
+
+# We need a separate thread to wait for notifications, 
+# but calling peripheral's methods from different threads creates issues,
+# so we will wrap all the calls into a thread
+class BluepyThreadedPeripheral(object):
+    def __init__(self, addr, addrType, controller):
+        self._call_queue = queue.Queue()
+        self._addr = addr
+        self._addrType = addrType
+        self._iface_number = _get_iface_number(controller)
+                
+        self._dispatcher_thread = Thread(target=self._dispatch_calls)
+        self._dispatcher_thread.setDaemon(True)
+        self._dispatcher_thread.setName("Bluepy call dispatcher")
+        self._dispatcher_thread.start()
+
+    def _dispatch_calls(self):
+        self._peripheral = btle.Peripheral(self._addr, self._addrType, self._iface_number)
+
+        while True:
+            try:
+                try:
+                    method = self._call_queue.get(False)
+                    method()
+                except queue.Empty:
+                    pass
+                self._peripheral.waitForNotifications(1.)
+            except Exception as ex:
+                log.exception('Exception in call dispatcher thread', exc_info=ex)
+                if KILL_ON_DISPATCHER_EXCEPTION:
+                    exit(1)
+
+    def write(self, handle, data):
+        self._call_queue.put(lambda: self._peripheral.writeCharacteristic(handle, data))
+
+    def set_notify_handler(self, handler):
+        delegate = BluepyDelegate(handler)
+        self._call_queue.put(lambda: self._peripheral.withDelegate(delegate))
+
+    def disconnect(self):
+        self._call_queue.put(lambda: self._peripheral.disconnect())
+
+
+class BluepyConnection(Connection):
+    def __init__(self, controller='hci0'):
+        Connection.__init__(self)
+        self._peripheral = None  # :type BluepyThreadedPeripheral
+        self._controller = controller
+
+    def connect(self, hub_mac=None):
+        log.debug("Trying to connect client to MoveHub with MAC: %s", hub_mac)
+        scanner = btle.Scanner()
+
+        while not self._peripheral:
+            log.info("Discovering devices...")
+            scanner.scan(1)
+            devices = scanner.getDevices()
+
+            for dev in devices:
+                address = dev.addr
+                addressType = dev.addrType
+                name = dev.getValueText(COMPLETE_LOCAL_NAME_ADTYPE)
+                log.debug("Found dev, name: {}, address: {}".format(name, address))
+
+                if name == LEGO_MOVE_HUB or hub_mac == address:
```

Comment:
There was a bug fixed around this logic, please look at https://github.com/undera/pylgbst/blob/master/pylgbst/comms/cgatt.py#L103
and modify this line to match it.

Reply:
Ok, I will check it out later today